modules/space/horizonsfile.h

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * A Horizons file is a text file generated from NASA JPL HORIZONS Website
+ * (https://ssd.jpl.nasa.gov/horizons.cgi). The implementation supports both Vector
+ * and Observer as Horizons data table
+ *
+ * In case of Vector table data the implementation expects a file with format:
+ * TIME(JulianDayNumber = A.D. YYYY-MM-DD HH:MM:SS TDB)
+ *   X(km) Y(km) Z(km)
+ * TIME - Only the "YYYY-MM-DD HH:MM:SS" part is of interest, the rest is ignored
+ * X - X position in kilometers in Ecliptic J2000 reference frame
+ * Y - Y position in kilometers in Ecliptic J2000 reference frame
+ * Z - Z position in kilometers in Ecliptic J2000 reference frame
+ * Changes required in the "Table Settings" for compatible data:
+ * 1. Under "Select Output Quantities" choose option "Position components {x, y, z} only"
+ * 2. Uncheck the "Vector labels" options
+ *
+ * In case of Observer table data the implementation expects a file with format:
+ * TIME(YYYY-MM-DD HH:MM:SS) Range(km) GalLon(degrees) GalLat(degrees)
+ * Range - The distance from target to observer in kilometers
+ * GalLon - Galactic Longitude in degrees
+ * GalLat - Galactic Latitude in degrees
+ * Changes required in the "Table Settings" for compatible data:
+ * 1. Under "Observer Table Settings" uncheck all options except
+ *    "Observer range & range-rate" and "Galactic longitude & latitude"
+ * 2. Change "Range units" to "kilometers (km)" instead of "astronomical units (au)"
+ * 3. Check the "Suppress range-rate" option
+ */
+class HorizonsFile {
+public:
+    enum class ResultCode {
+        Valid,
+        InvalidFormat,
+        Empty,
+
+        // Erros caught by the error field in the json output
+        ErrorSize,
+        ErrorSpan,
+        ErrorTimeRange,
+        ErrorNoObserver,
+        ErrorObserverTargetSame,
+        ErrorNoData,
+        MultipleObserverStations,
+
+        // Erros/problems NOT caught by the error field in the json output
+        MultipleObserver,
+        ErrorNoTarget,
+        MultipleTarget,
+
+        UnknownError
+    };
+
+    enum class Type {
+        Observer, // Default
+        Vector,   // Default for sending for new data
+        Invalid   // If errors or empty etc
+    };
+
+    struct HorizonsKeyframe {
+        double time;            // J2000 seconds
+        glm::dvec3 position;    // GALACTIC cartesian coordinates in meters
+    };
+
+    struct HorizonsResult {
+
+        HorizonsResult()
+            : type(Type::Invalid), errorCode(ResultCode::UnknownError)
+        {}
+
+        Type type;
+        ResultCode errorCode;
+        std::vector<HorizonsKeyframe> data;
+    };
+
+    HorizonsFile();
+    HorizonsFile(std::filesystem::path file);
+    HorizonsFile(std::filesystem::path filePath, const std::string& result);
+
+    void setFile(std::filesystem::path file);
+    const std::filesystem::path& file() const;
+    std::filesystem::path& file();
+
+    static std::string constructUrl(Type type, const std::string& target,
```

Comment:
This function doesn't have to be a member function of the HorizonsFile at all and can just be a stand-alone function instead